Add RPCErrors to more QueryService RPC responses

aaijazi opened the pull request:
@alainjobart 

(this includes #760 in it, so you can skip the first commit)

I switched over everything except for StreamExecute. I'm actually not sure how we're going to be able to pass the error back in the RPC response for that streaming queries. We should discuss in person on Monday...

I'm also unsure if my changes to Commit and Rollback are safe, please confirm. It looks like rpc.Unused{} in the server signature would have been safe to change, but I'm not sure about the noOutput string pointer. The clients were sending rpc.Unused{} though, so it might be okay.

aaijazi commented:
Turns out that this won't be safe until after #771 is rolled out everywhere.

aaijazi commented:
@alainjobart, PTAL, made the following changes:

1. Added RPCErrors to StreamExecute, handled it in the server and go/python clients. Also added tests to prove that it works. 

Sending errors in this way (as a last, additional result on the channel) was what sougou and I came up with yesterday. Let me know if there are objections or other suggestions. 

2. Added new RPC end-points for the new Commit and Rollback behavior, as they are not safe to change in-place. I don't expect anything to use the new end-points, but they're useful to keep around for testing.

alainjobart commented:
@guoliang100 will need a new begin, commit and rollback to pass in the callerID, maybe you guys should sync up and make a compatible change.

Also, if you're gonna change these guys to a new endpoint, you should make them match the new proto3 interface that's now defined in proto/, that'll be less code later to change. I mean introduce BeginRequest and BeginResponse for instance...

(rest of the change looks good)

alainjobart commented on test/vtdb_test.py:
```diff
@@ -375,6 +398,38 @@ def test_tablet_restart_stream_execute(self):
+    This is for testing app failures, not infrastructure failures. We simulate
+    an app failure by killing MySQL on the master tablet.
+
+    This test is inherently racy, as it tries to kill MySQL in the middle
```
One way to make it less racy is to execute a sleep(X) where X is big. It doesn't have the same effect as no result is returned, but it interrupts a query in the middle.

aaijazi replied:
sleep(X) on the mysql side? I'm trying to interrupt a streaming query after at least the first QueryResult is returned. I'm not sure sleeping MySQL is any less racy.

alainjobart replied:
I meant StreamExecute("select sleep(30)") or whatever the syntax is for 30 seconds. You'll get the Fields, then nothing for 30 seconds. If you sleep for 10s and then kill mysqld, should not be racy at all.

aaijazi replied:
Ah, makes sense. Tried this out, but it turns out that our sqlparser doesn't like "select sleep(30)", so I'm going to leave it as is.

(It did help me discover a bug though)

aaijazi replied:
I killed this test after I found [queryservice/stream_tests/test_streaming_terminate](https://github.com/youtube/vitess/blob/master/test/queryservice_tests/stream_tests.py#L102), which does the same thing but without destructively killing MySQL.

aaijazi commented:
@alainjobart @guoliang100, PTAL. I'd like to merge tomorrow morning...

(apologies for the PR that has grown in size, I'll try to do fewer end-points at a time for the next ones)

alainjobart commented:
LGTM but I'd like Liang to review it too, since it contains a bunch of code he's gonna need. Thanks!

guoliang100 commented on go/vt/tabletserver/gorpcqueryservice/sqlquery.go:
```diff
@@ -23,13 +23,40 @@ type SqlQuery struct {
 }
 
 // Begin is exposing tabletserver.SqlQuery.Begin
 func (sq *SqlQuery) Begin(ctx context.Context, session *proto.Session, txInfo *proto.TransactionInfo) (err error) {
```
We also need new "Begin" RPC accepting BeginRequest and returning BeginResponse.

aaijazi replied:
Done.

guoliang100 commented on go/vt/tabletserver/gorpcqueryservice/sqlquery.go:
```diff
@@ -23,13 +23,40 @@ type SqlQuery struct {
+// UnsupportedNewCommit should not be used by anything other than tests.
+// It will eventually replace Commit, but it breaks compatibility with older clients.
+// Once all clients are upgraded, it can be replaced.
+func (sq *SqlQuery) UnsupportedNewCommit(ctx context.Context, commitRequest *proto.CommitRequest, commitResponse *proto.CommitResponse) (err error) {
```
We can do a 2-step migration to the new RPCs (begin, commit, rollback).
It is better to name them Begin2/Commit2/Rollback2

guoliang100 commented on go/vt/tabletserver/gorpcqueryservice/sqlquery.go:
```diff
@@ -47,32 +91,57 @@ func (sq *SqlQuery) Rollback(ctx context.Context, session *proto.Session, noOutp
+	if *tabletserver.RPCErrorOnlyInReply {
+		// If there was an app error, send a QueryResult back with it.
+		qr := new(mproto.QueryResult)
+		tabletserver.AddTabletErrorToQueryResult(tErr, qr)
```
It will be tricky to roll this out. The client needs to be updated first to handle both cases. However, rdonly clients may take a much longer to fully updated.
I think the StreamExecute call only returns error if it cannot start the streaming. Maybe you can modify the underlying function, and see if the error can be attached to one of existing message (and existing client could handle it).

aaijazi replied:
Unfortunately, there isn't any way to attach an error to an existing message. Each message is sent as the DB streams the results back: https://github.com/youtube/vitess/blob/d4a89685afa4f3b60af38805e0f3d9392478d539/go/vt/dbconnpool/connection.go#L84

If a row is successfully received, it is sent. If an error is received, no more messages are sent, and an error is returned instead.

Basically, by the time an error happens, every message that was going to be sent has already been sent. That's why we need to send one last message (as I'm doing here) to send the error.

I think the alternative way to do it would be to use the API that Alain has for his proto3 stuff - return a StreamExecuteResponse that would contain a QueryResult and an Err, and clients would unwrap that to figure out what to do. In practice, that would look pretty much the same as what's happening here, but backwards compatibility would be maintained by adding an new end-point.

guoliang100 commented on go/vt/tabletserver/gorpcqueryservice/sqlquery.go:
```diff
@@ -23,13 +23,40 @@ type SqlQuery struct {
-	return sq.server.GetSessionId(sessionParams, sessionInfo)
+	tErr := sq.server.GetSessionId(sessionParams, sessionInfo)
+	tabletserver.AddTabletErrorToSessionInfo(tErr, sessionInfo)
+	if *tabletserver.RPCErrorOnlyInReply {
```
If you can solve the StreamExecute case, you do not need this flag.
-> You will not be able to switch this flag without causing compatibility issue (when you do, instead, you can just delete the obsolete code).

aaijazi replied:
I'd still prefer to keep the flag, because it's safer/quicker/more easily revertible than just deleting the obsolete code. Once the flag is flipped everywhere, we can go in and delete the unneeded  code (and the flag, eventually).

guoliang100 commented on go/vt/tabletserver/tablet_error.go:
```diff
@@ -212,28 +213,89 @@ func logError(queryServiceStats *QueryServiceStats) {
+
+// AddTabletErrorToQueryResult will mutate a QueryResult struct to fill in the Err
+// field with details from the TabletError.
+func AddTabletErrorToQueryResult(err error, reply *mproto.QueryResult) {
```
Since rpcErrFromTabletError() checks if err == nil, I think you do not need these functions.
The references to these functions can be written as:
reply.Err = rpcErrFromTabletError(err)

aaijazi replied:
Not exactly true. The main reason that I want these different functions is because I want strong typing, so I need one function per type. (the alternative would be to specify `interface` in the signature, which I'd prefer to avoid)

The reason that I have the nil check in each function is that, if err is nil, I don't want to overwrite the existing reply.Err with nil. The current behavior will ignore reply.Err when err is nil.

guoliang100 commented on go/vt/tabletserver/tabletconn/tablet_conn.go:
```diff
@@ -76,6 +76,11 @@ type TabletConn interface {
 
+	// These should not be used for anything except tests for now; they will eventually
+	// replace the existing methods.
+	UnsupportedNewCommit(context context.Context, transactionId int64) error
```
It is better to name these as Commit2, and leave it untouched in future.
Otherwise, you need to modify the interface again.

aaijazi replied:
Done.

aaijazi commented:
Addressed comments. @guoliang100 , do you think it'd be worth it to add another StreamExecute end-point, so that we don't break backwards compatibility for older clients?

guoliang100 commented:
I think it is worth to add another RPC.
Alternatively, we need to make sure all old clients are updated, before rolling this change out.

aaijazi commented:
Added StreamExecute2. The clients for both StreamExecute and StreamExecute2 should be compatible, but the StreamExecute server end-point will never return an error in the response. This protects old clients even after the flag is flipped.

aaijazi commented:
This is the final version, planning on merging this soon.
